pkg/koordlet/prediction/prediction.go

```diff
@@ -0,0 +1,181 @@
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/metriccache"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/statesinformer"
+)
+
+type UIDType string
+
+type UIDGenerator interface {
+	Pod(pod *v1.Pod) UIDType
+	Node(node *v1.Node) UIDType
+}
+
+type Options struct {
+	Filepath string
+	// TODO add configs here
+}
+
+// The kubernetes UID is unique within the POd lifecycle, so use this first. If there
+// are some special scenarios in the future, such as deleting a Pod and creating a Pod
+// with the same name, consider using NamespacedName as the UID.
+type generator struct {
+}
+
+func (gen *generator) Pod(pod *v1.Pod) UIDType {
+	return UIDType(pod.GetUID())
+}
+
+func (gen *generator) Node(node *v1.Node) UIDType {
+	return UIDType(node.GetUID())
+}
+
+type Result struct {
+	// Use different quantile type as key, currently support "p60", "p90", "p95" "p98", "max".
+	Data map[string]v1.ResourceList
+}
+
+// FIXME
+// This is used for the agent's dependence on the basic data structure, and the
+// basic data structure will be reconstructed later to better support testing.
+type Informer interface {
+	HasSynced() bool
+	ListPods() []*v1.Pod
+	GetNode() *v1.Node
+}
+
+func NewInformer(statesInformer statesinformer.StatesInformer) Informer {
+	return &informer{
+		statesInformer: statesInformer,
+	}
+}
+
+type informer struct {
+	statesInformer statesinformer.StatesInformer
+}
+
+func (i *informer) HasSynced() bool {
+	return i.statesInformer.HasSynced()
+}
+
+func (i *informer) ListPods() []*v1.Pod {
+	pods := i.statesInformer.GetAllPods()
+	result := make([]*v1.Pod, len(pods))
+	for i := range pods {
+		result[i] = pods[i].Pod
+	}
+	return result
+}
+
+func (i *informer) GetNode() *v1.Node {
+	return i.statesInformer.GetNode()
+}
+
+type MetricDesc struct {
+	UID UIDType
+}
+
+type MetricKey int
+
+const (
+	CPUUsage MetricKey = iota
+	MemoryUsage
+)
+
+type MetricServer interface {
+	GetPodMetric(desc MetricDesc, m MetricKey) (float64, error)
+	GetNodeMetric(desc MetricDesc, m MetricKey) (float64, error)
+}
+
+func NewMetricServer(metricCache metriccache.MetricCache) MetricServer {
+	return &metricServer{
+		metricCache: metricCache,
+	}
+}
+
+type metricServer struct {
+	metricCache metriccache.MetricCache
+}
+
+func (ms *metricServer) GetPodMetric(desc MetricDesc, m MetricKey) (float64, error) {
+	now := time.Now()
+	start := now.Add(-2 * DefaultTrainingInterval)
+	querier, err := ms.metricCache.Querier(start, now)
+	if err != nil {
+		return 0, err
+	}
+
+	podProperties := metriccache.MetricPropertiesFunc.Pod(string(desc.UID))
+	queryPodMetric := func(m metriccache.MetricResource) (float64, error) {
+		meta, err := m.BuildQueryMeta(podProperties)
+		if err != nil {
+			return 0, err
+		}
+		result := metriccache.DefaultAggregateResultFactory.New(meta)
+		if err = querier.Query(meta, nil, result); err != nil {
+			return 0, err
+		}
+		return result.Value(metriccache.AggregationTypeLast)
```

Comment:
This adds the last metric point in the last 2 mins. It may be inaccurate when the pod usage varies in seconds.
How about using the avg/max/percentile metric and zooming in the window?